### PR TITLE
test: [TECH-887] improve rules engine unit test coverage

### PR DESCRIPTION
**Pull request by simonadomnisoru**
[TECH-887](https://dhis2.atlassian.net/browse/TECH-887)

**Tech summary**
- Improved the unit tests coverage for:
    - `packages/rules-engine`
    
    packages:rules-engine

    - `src/core_modules/capture-core/rules` folder
    
    capture-core:rules

- The first commit is just a format of the test files. Therefore, I recommend that you take a look at the https://github.com/dhis2/capture-app/pull/3462/commits/296c1be84a92a219f1b13c53ce4a25188a41739b commit to visualize the changes more easily.

[TECH-887]: https://dhis2.atlassian.net/browse/TECH-887?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by eirikhaugstulen** [comment]
Hey @simonadomnisoru!
This is great and I think the test coverage improvements speak for itself.

I do however get a lot of warnings and actually some errors as well when running the tests locally.

image

These errors and warnings are not part of the tests we have in master. Is this something we need to address or is it fine as is?

**Comment by simonadomnisoru** [comment]
> I do however get a lot of warnings and actually some errors as well when running the tests locally.
> 
> These errors and warnings are not part of the tests we have in master. Is this something we need to address or is it fine as is?

Hey @eirikhaugstulen,

Warning and error logs are expected in the console. They are not shown in master because so far we mainly tested the "happy path" In this PR, I added a bunch of test cases where I deliberately tested with invalid data to make sure the input was properly validated and the expected errors were thrown. Let me know if you have any more questions.

Thanks for the review!

**Review by eirikhaugstulen — APPROVED** [review]
That makes sense!
I think this change looks good, Simona! 👍

**Review by superskip — COMMENTED** [review]
Hi @simonadomnisoru! I only have two comments about some changes you made in the rules engine. I haven't looked at the tests you added, mostly because it seems like a frightfully big task to get the overview of what's been implemented and then try to figure out whether the current coverage is satisfactory. It might help with a little more information on the strategy you used to get to this point.

File: packages/rules-engine/src/services/VariableService/helpers/structureEvents.js [hunk 1]
```diff
@@ -38,7 +38,7 @@ export const getStructureEvents = (compareDates: CompareDates) => {
                     event.eventId !== currentEvent.eventId,
             );
 
-        const events = [...otherEventsFiltered, currentEvent]
```
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply packages/rules-engine/src/services/VariableService/helpers/structureEvents.js:41]: Good catch! Suggestion implemented. Thanks for the feedback! 😄

[the remark this answers is not in the document]
Comment (superskip) [reply packages/rules-engine/src/services/VariableService/helpers/structureEvents.js:41]: Ok.. what about `eventId`? :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (superskip) [reply packages/rules-engine/src/services/VariableService/helpers/structureEvents.js:41]: Maybe it does not exist during registration?

[the remark this answers is not in the document]
Comment (superskip) [reply packages/rules-engine/src/services/VariableService/helpers/structureEvents.js:41]: No, you're right :face_exhaling:
Guess it has to be this way :+1:

**Pull request merged**